pkg/pool-utils/test/RebalancedBasePool.test.ts

```diff
@@ -0,0 +1,198 @@
+
+describe('RebalancedBasePool', function () {
+  let user: SignerWithAddress;
+  let vault: Vault, relayer: Contract, pool: Contract, tokens: TokenList, assetManagers: Contract[];
+
+  before('set signer', async () => {
+    [, user] = await ethers.getSigners();
+  });
+
+  sharedBeforeEach('deploy pool', async () => {
+    vault = await Vault.create();
+    relayer = await deploy('MockRebalancedBasePoolRelayer');
+    tokens = await TokenList.create(['DAI', 'MKR', 'SNX'], { sorted: true });
+    await tokens.mint({ to: user, amount: fp(50) });
+    await tokens.approve({ to: vault.instance, amount: fp(1000), from: user });
+
+    assetManagers = [
+      await deploy('v2-asset-manager-utils/MockAssetManager', { args: [tokens.get(0).address] }),
+      await deploy('v2-asset-manager-utils/MockAssetManager', { args: [tokens.get(1).address] }),
+      await deploy('v2-asset-manager-utils/MockAssetManager', { args: [tokens.get(2).address] }),
+    ];
+
+    pool = await deploy('MockRebalancedBasePool', {
+      args: [
+        vault.address,
+        GeneralPool,
+        'BPT',
+        'BPT',
+        tokens.addresses,
+        assetManagers.map((assetManager) => assetManager.address),
+        fp(0.1),
+        0,
+        0,
+        relayer.address,
+        ZERO_ADDRESS,
+      ],
+    });
+  });
+
+  describe('relayer', () => {
+    it('uses the given relayer', async () => {
+      expect(await pool.getRelayer()).to.be.equal(relayer.address);
+    });
+  });
+
+  const join = async () => {
+    return vault.joinPool({
+      poolAddress: pool.address,
+      poolId: await pool.getPoolId(),
+      recipient: user.address,
+      currentBalances: Array(tokens.length).fill(fp(1000)),
+      tokens: tokens.addresses,
+      lastChangeBlock: 0,
+      protocolFeePercentage: 0,
+      data: '0x',
+      from: user,
+    });
+  };
+
+  describe('join', () => {
+    context('when no asset manager requires a rebalance', () => {
+      sharedBeforeEach('mock asset manager', async () => {
+        await Promise.all(assetManagers.map((assetManager) => assetManager.mockShouldRebalance(false)));
+      });
+
+      context('when the relayer tells it has not called the pool', () => {
+        sharedBeforeEach('mock relayer', async () => {
+          await relayer.mockHasCalledPool(false);
+        });
+
+        it('does not revert', async () => {
+          await expect(join()).not.to.be.reverted;
+        });
+      });
+
+      context('when the relayer tells it has called the pool', () => {
+        sharedBeforeEach('mock relayer', async () => {
+          await relayer.mockHasCalledPool(true);
+        });
+
+        it('does not revert', async () => {
+          await expect(join()).not.to.be.reverted;
+        });
+      });
+    });
+
+    context('when some asset managers require a rebalance', () => {
+      sharedBeforeEach('mock asset manager', async () => {
+        await assetManagers[1].mockShouldRebalance(true);
+      });
```

Comment:
This suggestion might be a bit pedantic and overkill, so take it with a grain of salt.

It'd be _great_ if we had not just this scenario, but actually programmatically iterated over the asset managers for `tokens[0..7]`, and tested that when _any_ of all 8 are unbalanced, the checks occur. We're otherwise leaving a large portion of the error-prone `if-else` chain in `_ensureAssetManagerBalanced` untested.

Similarly, it'd be good to check that each asset manager receives the expected cash balances, but that might be a tad complicated as we can't e.g. emit events in the `view` call.